cli_tools/common/utils/storage/buffered_writer.go

```diff
@@ -0,0 +1,236 @@
+//  Copyright 2020 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package storage
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"math"
+	"os"
+	"path"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/domain"
+	pathutils "github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/path"
+	"google.golang.org/api/googleapi"
+)
+
+var gcsPermissionErrorRegExp = regexp.MustCompile(".*does not have storage.objects.create access to .*")
+
+type gcsClient func(ctx context.Context, oauth string) (domain.StorageClientInterface, error)
+
+// BufferedWriter is responsible for multipart component upload while using a local buffer.
+type BufferedWriter struct {
+	// These fields are read only.
+	cSize    int64
+	prefix   string
+	ctx      context.Context
+	oauth    string
+	client   gcsClient
+	id       string
+	bkt, obj string
+
+	upload    chan string
+	tmpObjs   []string
+	tmpObjsMx sync.Mutex
+
+	sync.Mutex
+	sync.WaitGroup
+	bytes int64
+	part  int
+	file  *os.File
+}
+
+// NewBuffer creates a BufferedWriter
+func NewBuffer(ctx context.Context, size, workers int64, client gcsClient, oauth, prefix, bkt, obj string) *BufferedWriter {
```

Comment:
When I see 'NewBuffer', I think of this from the stdlib: https://golang.org/pkg/bytes/#example_Buffer ; is that your intent? Any reason to not use 'NewBufferedWriter'?

Reply:
I reused the method name when I refactored. I think NewBufferedWriter makes sense, I will change it. Thanks :)